### DISCUSSION TITLE
Is hydra the right tool?

### DISCUSSION THREAD

**Discussion by softshipper** [ory/hydra#2484] [Q&A] [answered]
Hi all  

I am going to build a web application that allows users to sign in with their Google or Twitter account. I think OpenID Connect(OAuth2) is the standard today to verify the identity.
I also want to provide several API services that can be only accessed with a valid access token from either Google or Twitter. 

[![enter image description here][1]][1]

For example, all the four API's above are going to be public and so I have to protect from unauthorized users. For NodeJS based API services I can use http://www.passportjs.org/ to protect all APIs.

Assume, in the future the number of API's will be grow for example up to 20 API's and sign in with Facebook account will be also allowed. 
Again, all the API's have to be protected and I have to do it 16 times with http://www.passportjs.org/. 
In addition add the new provider Facebook, I have to do the changes on all 20 APIs.

[![enter image description here][2]][2]

The question is, is their a way to keep centralized, which means in the future when I will provide more the providers for example GITHUB for sign in I would like to do changes in one place not in 20 places.
Is the tool https://www.ory.sh/hydra what I need?

  [1]: https://i.stack.imgur.com/at6at.png
  [2]: https://i.stack.imgur.com/H3Vg0.png

Thanks

**Comment by vinckr** [comment, marked as answer]
In short: 
You are looking for [Ory Kratos](https://github.com/ory/kratos/).
It will enable you to easily add new OIDC providers. 

Also check out this article, if you haven't already: [Do You Need OAuth2?](https://www.ory.sh/hydra/docs/concepts/before-oauth2)

I didn't quite get your second use case: 
>can be only accessed with a valid access token from either Google or Twitter.
So people could add your APIs to their account and then generate PATs to access them?

**Reply by softshipper** [reply]
In short, what I am trying to achieve is to protect my APIs again unauthorized users. However, I do not plan to provide sign up just only sign in with google, facebook etc. So is kratos the right choice?

**Reply by vinckr** [reply]
To protect your APIs from unauthorized requests you probably want something like a reverse proxy, right? 
([Ory Oathkeepe](https://www.ory.sh/oathkeeper/docs/#reverse-proxy)r has that functionality).


>  I do not plan to provide sign up just only sign in with google, facebook etc.  

You can specify this in Kratos, so for example omit the "normal" signup (email/pw) and only enable social sign-ins. 
Basically you can form the login flow & UI how it suits you. 

I believe you need something to manage users, even if they only use social sign-in and Hydra does not offer that functionality.  You also would need to store the access/refresh tokens somewhere. 

Hydra would be more if "you want developers you don't know to interact with your users and your users' data", 
so for example if your users log into a third-party application and call your API through that application.

I hope that makes some sense and feel free to correct me if I got something wrong!

**Comment by vinckr** [comment]
@softshipper 
The Ory Kratos and Ory Hydra integration just landed with Hydra 2.0, 
see the [release notes](https://github.com/ory/hydra/releases/tag/v2.0.0) for more details!
